// superplate-cli/lib/Helper/download/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DownloadHelper = void 0;

var _temp = require("temp");

var _got = _interopRequireDefault(require("got"));

var _tar = _interopRequireDefault(require("tar"));

var _fs = require("fs");

var _stream = require("stream");

var _parseGithubUrl = _interopRequireDefault(require("parse-github-url"));

var _util = require("util");

var _path = require("path");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const pipeline = (0, _util.promisify)(_stream.Stream.pipeline);
const TEMP_PREFIX = "superplate-core-plugins.temp";
const DownloadHelper = {
  DownloadAndGetPath: async path => {
    try {
      const tempFolder = (0, _temp.mkdirSync)({
        dir: process.cwd(),
        prefix: "."
      });
      const tempFile = (0, _path.join)(tempFolder, `${TEMP_PREFIX}-${Date.now()}`);
      const {
        owner,
        name,
        branch
      } = (0, _parseGithubUrl.default)(path);
      const url = `https://codeload.github.com/${owner}/${name}/tar.gz/${branch}`;
      await pipeline(_got.default.stream(url), (0, _fs.createWriteStream)(tempFile));
      await _tar.default.x({
        file: tempFile,
        cwd: tempFolder,
        strip: 1
      });
      return tempFolder;
    } catch (e) {
      throw new Error(e instanceof Error ? e.message : e);
    }
  }
};
exports.DownloadHelper = DownloadHelper;
