// superplate-cli/lib/Helper/plugin/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handleIgnore = exports.concatExtend = exports.getExtend = exports.getPluginsArray = exports.extendBase = void 0;

var _path = _interopRequireDefault(require("path"));

var _ = require("./..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const extendBase = {
  _app: {
    import: [],
    inner: [],
    wrapper: []
  },
  _document: {
    import: [],
    initialProps: []
  }
};
exports.extendBase = extendBase;

const getPluginsArray = answers => {
  return Object.entries(answers).reduce((acc, [key, value]) => {
    if (typeof value === "boolean" && value) return [...acc, key];
    if (typeof value === "string") return [...acc, value];
    if (Array.isArray(value)) return [...acc, ...value];
    return acc;
  }, []).filter(value => value !== "none" && value !== "no");
};

exports.getPluginsArray = getPluginsArray;

const getExtend = (pluginPath, pluginName) => {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const pluginExtend = require(_path.default.join(pluginPath, "plugins", pluginName, "extend.js"));

    return pluginExtend;
  } catch (e) {
    return undefined;
  }
};

exports.getExtend = getExtend;

const concatExtend = (base, plugins, sourcePath, answers) => {
  const merged = (0, _.mergeWithUnionArray)(base, ...plugins.map(plugin => {
    const pluginExtendFile = getExtend(sourcePath, plugin);

    if (pluginExtendFile) {
      const pluginExtends = pluginExtendFile.extend(answers);
      return pluginExtends;
    }

    return {};
  }));
  return merged;
};

exports.concatExtend = concatExtend;

const handleIgnore = (ignores, answers, plugin) => {
  const filters = {};
  ignores.forEach(ignore => {
    if (!!ignore.plugin === false || !!ignore.plugin && ignore.plugin.includes(plugin)) {
      var _ignore$when;

      const condition = (_ignore$when = ignore.when) === null || _ignore$when === void 0 ? void 0 : _ignore$when.call(ignore, answers);

      if (condition) {
        ignore.pattern.forEach(pattern => {
          filters[pattern] = false;
        });
      }
    }
  });
  return filters;
};

exports.handleIgnore = handleIgnore;
