// superplate-cli/lib/Helper/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "mergePackages", {
  enumerable: true,
  get: function () {
    return _merge.mergePackages;
  }
});
Object.defineProperty(exports, "mergeJSONFiles", {
  enumerable: true,
  get: function () {
    return _merge.mergeJSONFiles;
  }
});
Object.defineProperty(exports, "mergeBabel", {
  enumerable: true,
  get: function () {
    return _merge.mergeBabel;
  }
});
Object.defineProperty(exports, "mergePluginData", {
  enumerable: true,
  get: function () {
    return _merge.mergePluginData;
  }
});
Object.defineProperty(exports, "mergeWithUnionArray", {
  enumerable: true,
  get: function () {
    return _merge.mergeWithUnionArray;
  }
});
Object.defineProperty(exports, "extendBase", {
  enumerable: true,
  get: function () {
    return _plugin.extendBase;
  }
});
Object.defineProperty(exports, "getPluginsArray", {
  enumerable: true,
  get: function () {
    return _plugin.getPluginsArray;
  }
});
Object.defineProperty(exports, "getExtend", {
  enumerable: true,
  get: function () {
    return _plugin.getExtend;
  }
});
Object.defineProperty(exports, "concatExtend", {
  enumerable: true,
  get: function () {
    return _plugin.concatExtend;
  }
});
Object.defineProperty(exports, "handleIgnore", {
  enumerable: true,
  get: function () {
    return _plugin.handleIgnore;
  }
});
Object.defineProperty(exports, "get_source", {
  enumerable: true,
  get: function () {
    return _source.get_source;
  }
});
Object.defineProperty(exports, "get_project_types", {
  enumerable: true,
  get: function () {
    return _source.get_project_types;
  }
});
Object.defineProperty(exports, "is_multi_type", {
  enumerable: true,
  get: function () {
    return _source.is_multi_type;
  }
});
Object.defineProperty(exports, "prompt_project_types", {
  enumerable: true,
  get: function () {
    return _source.prompt_project_types;
  }
});
Object.defineProperty(exports, "UrlHelper", {
  enumerable: true,
  get: function () {
    return _url.UrlHelper;
  }
});
Object.defineProperty(exports, "GitHelper", {
  enumerable: true,
  get: function () {
    return _git.GitHelper;
  }
});
Object.defineProperty(exports, "FSHelper", {
  enumerable: true,
  get: function () {
    return _fs.FSHelper;
  }
});
Object.defineProperty(exports, "tips", {
  enumerable: true,
  get: function () {
    return _tips.tips;
  }
});
Object.defineProperty(exports, "BinaryHelper", {
  enumerable: true,
  get: function () {
    return _binary.BinaryHelper;
  }
});
Object.defineProperty(exports, "get_presets", {
  enumerable: true,
  get: function () {
    return _preset.get_presets;
  }
});
Object.defineProperty(exports, "get_prompts_and_choices", {
  enumerable: true,
  get: function () {
    return _lucky.get_prompts_and_choices;
  }
});
Object.defineProperty(exports, "get_random_answer", {
  enumerable: true,
  get: function () {
    return _lucky.get_random_answer;
  }
});
Object.defineProperty(exports, "get_random_answers", {
  enumerable: true,
  get: function () {
    return _lucky.get_random_answers;
  }
});
Object.defineProperty(exports, "prompt_telemetry", {
  enumerable: true,
  get: function () {
    return _telemetry.prompt_telemetry;
  }
});
Object.defineProperty(exports, "get_potential_package_managers", {
  enumerable: true,
  get: function () {
    return _npm.get_potential_package_managers;
  }
});
Object.defineProperty(exports, "HumanizeChoices", {
  enumerable: true,
  get: function () {
    return _humanize.HumanizeChoices;
  }
});
Object.defineProperty(exports, "getRandomName", {
  enumerable: true,
  get: function () {
    return _getRandomName.getRandomName;
  }
});
Object.defineProperty(exports, "DownloadHelper", {
  enumerable: true,
  get: function () {
    return _download.DownloadHelper;
  }
});

var _merge = require("./merge");

var _plugin = require("./plugin");

var _source = require("./source");

var _url = require("./url");

var _git = require("./git");

var _fs = require("./fs");

var _tips = require("./tips");

var _binary = require("./binary");

var _preset = require("./preset");

var _lucky = require("./lucky");

var _telemetry = require("./telemetry");

var _npm = require("./npm");

var _humanize = require("./humanize");

var _getRandomName = require("./getRandomName");

var _download = require("./download");
